aten/src/ATen/Half.cpp

```diff
@@ -12,16 +8,91 @@ static_assert(std::is_standard_layout<Half>::value, "at::Half must be standard l
 
 namespace detail {
 
-float halfbits2float(unsigned short bits) {
-  float value;
-  TH_halfbits2float(&bits, &value);
-  return value;
+// Host functions for converting between FP32 and FP16 formats
+
+float halfbits2float(unsigned short h)
+{
+    unsigned sign = ((h >> 15) & 1);
+    unsigned exponent = ((h >> 10) & 0x1f);
+    unsigned mantissa = ((h & 0x3ff) << 13);
+
+    if (exponent == 0x1f) {  /* NaN or Inf */
+        mantissa = (mantissa ? (sign = 0, 0x7fffff) : 0);
+        exponent = 0xff;
+    } else if (!exponent) {  /* Denorm or Zero */
+        if (mantissa) {
+            unsigned int msb;
+            exponent = 0x71;
+            do {
+                msb = (mantissa & 0x400000);
+                mantissa <<= 1;  /* normalize */
+                --exponent;
+            } while (!msb);
+            mantissa &= 0x7fffff;  /* 1.mantissa is implicit */
+        }
+    } else {
+        exponent += 0x70;
+    }
+
+    unsigned result_bit = (sign << 31) | (exponent << 23) | mantissa;
+
+    // Reinterpret the result bit pattern as a float
+    float result_float;
+    std::memcpy(&result_float, &result_bit, sizeof(result_float));
```

Comment:
I think this should be possible with c++ static cast

Reply:
Actually it's reinterpret cast: https://en.cppreference.com/w/cpp/language/reinterpret_cast

Reply:
reinterpret_cast is less strict-aliasing safe than memcpy. Also, you can't reinterpret cast float to unsigned directly https://godbolt.org/g/DSBPnP

Reply:
TIL